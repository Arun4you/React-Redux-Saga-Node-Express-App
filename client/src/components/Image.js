import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import Avatar from '@material-ui/core/Avatar';
import Typography from '@material-ui/core/Typography';
import red from '@material-ui/core/colors/red';
import FavoriteIcon from '@material-ui/icons/Favorite';
import CommentIcon from '@material-ui/icons/Comment';
import Button from '@material-ui/core/Button';
import PropTypes from 'prop-types';

const styles = (theme) => ({
    card: {
        maxWidth: "auto",
        margin: 10,
    },
    media: {
        height: "100%",
        paddingTop: '56.25%', // 16:9
    },
    actions: {
        display: 'flex',
    },
    avatar: {
        backgroundColor: red[500],
    },
    button: {
        margin: theme.spacing.unit,
    }
});

const Image = ({ classes, image, handleLikes, redirectTo }) => {
    const { created, display_src, username, caption, charAt, likes, photo, _id } = image
    const id = photo || _id

    if (!display_src) {
        return null; 
    }

    return (
        <div onClick={()=>redirectTo(_id)} >
            <Card className={classes.card}>
                    <CardHeader
                        avatar={
                            <Avatar aria-label="ReactXImage" className={classes.avatar}>
                                {charAt}
                            </Avatar>
                        }
                        title={username}
                        subheader={created}
                    />
                    <CardActionArea>
                        <CardMedia
                            className={classes.media}
                            image={display_src || ' '}
                        />
                        <CardContent>
                            <Typography component="p">
                                {caption}
                            </Typography>
                        </CardContent>
                    </CardActionArea>

                <CardActions className={classes.actions} disableActionSpacing>
                    <Button data-test="favouriteButtonComponent" variant="outlined" color="primary" className={classes.button} id={id} onClick={handleLikes}>
                        <FavoriteIcon /> {likes}
                    </Button>
                    <Button data-test="commentButtonComponent" variant="outlined" color="primary" className={classes.button} id={id}>
                        <CommentIcon />
                    </Button>
                </CardActions>
            </Card>
        </div>
    );
}

Image.propTypes = {
    handleLikes: PropTypes.func,
    image: PropTypes.shape({
        created: PropTypes.string,
        display_src: PropTypes.string,
        username: PropTypes.string,
        caption: PropTypes.string,
        charAt: PropTypes.string,
        likes: PropTypes.number,
        photo: PropTypes.string,
        _id: PropTypes.string,
    })
}

export default withStyles(styles)(Image);
