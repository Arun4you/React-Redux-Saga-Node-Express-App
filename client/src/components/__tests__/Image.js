
import React from "react";
import Image from "./../Image";
import Button from '@material-ui/core/Button';
import { findByTestAtrr, findByAttr } from './../../../utils'
import { shallow, mount } from "enzyme";

describe(">>> <Image /> - SHOULD RENDER", () => {

    let wrapper;
    beforeEach(() => {
        const props = {
            image: {
                caption: "am watinggggg !! #Yummy",
                created: "2019-03-20T18:17:37.212Z",
                created_at: "2019-04-03T15:27:59.722Z",
                display_src: "https://picsum.photos/400/400?image=127",
                likes: 25,
                username: "Arun",
                __v: 0,
                _id: "5c9283c13bd4543164bd6507"
            },
            handleLikes: jest.fn(() => Promise.resolve())
        };
        wrapper = shallow(<Image {...props} />);
    });

    it("+++ SHOULD RENDER WITH PROPS", () => {
        expect(wrapper).toHaveLength(1)
    })

    // const component = findByTestAtrr(wrapper, Button);        
    // console.log(wrapper.instance().props)
    // expect(wrapper.find('Button')).toBe(1);
    // button.first().props().click()
    // console.log(button.first().props().handleLikes.calls.length)
    // expect(handleLikes.mock.calls.length).toBe(1);
})

describe(">>> <Image /> -  Testing Functionality ", () => {
    let wrapper
    const handleLikes = jest.fn()
    beforeEach(() => {
        const props = {
            image: {
                caption: "am watinggggg !! #Yummy",
                created: "2019-03-20T18:17:37.212Z",
                created_at: "2019-04-03T15:27:59.722Z",
                display_src: "https://picsum.photos/400/400?image=127",
                likes: 25,
                username: "Arun",
                __v: 0,
                _id: "5c9283c13bd4543164bd6507"
            },
            redirectTo: jest.fn(() => Promise.resolve()),
        };
        //Mount is used just to find Button Component inside Image Component
        wrapper = mount(<Image {...props} handleLikes={handleLikes} />);
    });

    it("--- SHOULD MATCH SNAPSHOT", () => {
        expect(wrapper).toMatchSnapshot();
    })

    it("+++ find two buttons ", () => {
        const button = wrapper.find(Button)
        // const button = findByTestAtrr(wrapper, 'favouriteButtonComponent')
        expect(button.length).toBe(2);
    })

    it("+++ click like buttons ", () => {        
        const button = wrapper.find(Button)
        button
            .first()
            .simulate('click')
        expect(handleLikes).toHaveBeenCalled();
    })
});

describe(">>> <Image /> - SHOULD NOT RENDER", () => {
    let wrapper
    beforeEach(() => {
        const props = {
            image: {
                caption: "am watinggggg !! #Yummy",
                created: "2019-03-20T18:17:37.212Z",
                created_at: "2019-04-03T15:27:59.722Z",
                likes: 25,
                username: "Arun",
                __v: 0,
                _id: "5c9283c13bd4543164bd6507"
            },
            handleLikes: jest.fn(() => Promise.resolve()),
            redirectTo: jest.fn(() => Promise.resolve())
        };
        wrapper = mount(<Image {...props} />);
    });

    it("--- props without image url", () => {
        const button = wrapper.find(Button)
        expect(button.length).toBe(0);
    })
})