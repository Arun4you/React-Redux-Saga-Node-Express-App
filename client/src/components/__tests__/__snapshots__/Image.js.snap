// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> <Image /> -  Testing Functionality  --- SHOULD MATCH SNAPSHOT 1`] = `
<div
  onClick={[Function]}
>
  <div
    className="MuiPaper-root-7 MuiPaper-elevation1-10 MuiPaper-rounded-8 MuiCard-root-6 Image-card-1"
  >
    <div
      className="MuiCardHeader-root-34"
    >
      <div
        className="MuiCardHeader-avatar-35"
      >
        <div
          aria-label="ReactXImage"
          className="MuiAvatar-root-40 MuiAvatar-colorDefault-41 Image-avatar-4"
        />
      </div>
      <div
        className="MuiCardHeader-content-37"
      >
        <span
          className="MuiTypography-root-43 MuiTypography-body2-51 MuiCardHeader-title-38"
        >
          Arun
        </span>
        <span
          className="MuiTypography-root-43 MuiTypography-body2-51 MuiTypography-colorTextSecondary-76 MuiCardHeader-subheader-39"
        >
          2019-03-20T18:17:37.212Z
        </span>
      </div>
    </div>
    <button
      className="MuiButtonBase-root-82 MuiCardActionArea-root-79"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <div
        className="MuiCardMedia-root-85 Image-media-2"
        style={
          Object {
            "backgroundImage": "url(\\"https://picsum.photos/400/400?image=127\\")",
          }
        }
      />
      <div
        className="MuiCardContent-root-87"
      >
        <p
          className="MuiTypography-root-43 MuiTypography-body1-52"
        >
          am watinggggg !! #Yummy
        </p>
      </div>
      <span
        className="MuiCardActionArea-focusHighlight-81"
      />
      <span
        className="MuiTouchRipple-root-126"
      />
    </button>
    <div
      className="MuiCardActions-root-88 MuiCardActions-disableActionSpacing-89 Image-actions-3"
    >
      <button
        className="MuiButtonBase-root-82 MuiButton-root-91 MuiButton-outlined-99 MuiButton-outlinedPrimary-100 Image-button-5"
        data-test="favouriteButtonComponent"
        disabled={false}
        id="5c9283c13bd4543164bd6507"
        onBlur={[Function]}
        onClick={[MockFunction]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-92"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-117"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
           
          25
        </span>
        <span
          className="MuiTouchRipple-root-126"
        />
      </button>
      <button
        className="MuiButtonBase-root-82 MuiButton-root-91 MuiButton-outlined-99 MuiButton-outlinedPrimary-100 Image-button-5"
        data-test="commentButtonComponent"
        disabled={false}
        id="5c9283c13bd4543164bd6507"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-92"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-117"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM18 14H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-126"
        />
      </button>
    </div>
  </div>
</div>
`;
